crates/ruff_workspace/src/options.rs

```diff
@@ -31,7 +31,7 @@ use ruff_python_formatter::{DocstringCodeLineWidth, QuoteStyle};
 
 use crate::settings::LineEnding;
 
-#[derive(Debug, PartialEq, Eq, Default, OptionsMetadata, Serialize, Deserialize)]
+#[derive(Clone, Debug, PartialEq, Eq, Default, OptionsMetadata, Serialize, Deserialize)]
```

Comment:
It's probably not a huge concern because options are rarely constructed, but where do we need to clone them now (cloning is somewhat expensive because the structure contains so many allocated sub-structures)?

Reply:
Yeah, I didn't like having to do this. It's because we initially use `clap` to parse `--config` arguments into variants of the `ConfigOption` enum, which is defined like this:

```rs
#[derive(Clone, Debug)]
pub enum ConfigOption {
    PathToConfigFile(PathBuf),
    ConfigOverride(Box<Options>),
}
```

One of `clap` or `serde` demands that the enum be clonable if we want to use it with the framework, and that means that `Options` needs to be clonable, which then cascades down into the fields of `Options`

Reply:
I see. Do you need to mutate the `Options` value? If not, using a `Rc` (or `Arc` if it needs to be threadsafe) might be an option which also gives you O(1) cloning

Reply:
> Do you need to mutate the `Options` value?

I don't think so, no!

> If not, using a `Rc` (or `Arc` if it needs to be threadsafe) might be an option which also gives you O(1) cloning

I'll give that a try

Reply:
Looks like it does need to be threadsafe, so `Rc` won't work for us here.

I tried doing this:

<details>
<summary>Diff I tried</summary>

```diff
--- a/crates/ruff/src/args.rs
+++ b/crates/ruff/src/args.rs
@@ -1,4 +1,5 @@
 use std::path::PathBuf;
+use std::sync::Arc;

 use anyhow::anyhow;
 use clap::builder::{TypedValueParser, ValueParserFactory};
@@ -691,7 +692,7 @@ fn resolve_bool_arg(yes: bool, no: bool) -> Option<bool> {
 #[derive(Clone, Debug)]
 pub enum ConfigOption {
     PathToConfigFile(PathBuf),
-    ConfigOverride(Box<Options>),
+    ConfigOverride(Arc<Options>),
 }

 #[derive(Clone)]
@@ -722,7 +723,7 @@ impl TypedValueParser for ConfigOptionParser {
             return Ok(ConfigOption::PathToConfigFile(path_to_config_file));
         }
         if let Ok(option) = toml::from_str(value) {
-            Ok(ConfigOption::ConfigOverride(option))
+            Ok(ConfigOption::ConfigOverride(Arc::new(option)))
         } else {
             let mut new_error =
                 clap::Error::new(clap::error::ErrorKind::ValueValidation).with_cmd(cmd);
diff --git a/crates/ruff_workspace/src/options.rs b/crates/ruff_workspace/src/options.rs
index 9eb54688a..8964d999b 100644
--- a/crates/ruff_workspace/src/options.rs
+++ b/crates/ruff_workspace/src/options.rs
@@ -31,7 +31,7 @@ use ruff_python_formatter::{DocstringCodeLineWidth, QuoteStyle};

 use crate::settings::LineEnding;

-#[derive(Clone, Debug, PartialEq, Eq, Default, OptionsMetadata, Serialize, Deserialize)]
+#[derive(Debug, PartialEq, Eq, Default, OptionsMetadata, Serialize, Deserialize)]
 #[serde(deny_unknown_fields, rename_all = "kebab-case")]
 #[cfg_attr(feature = "schemars", derive(schemars::JsonSchema))]
 pub struct Options {
```

</details>

But the compiler complains at me:

```
error[E0507]: cannot move out of an `Arc`
   --> crates\ruff\src\args.rs:539:53
    |
539 |                         Configuration::from_options(*overridden_option, &path_dedot::CWD)?,
    |                                                     ^^^^^^^^^^^^^^^^^^ move occurs because value has type `ruff_workspace::options::Options`, which does not implement the `Copy` trait
```

It seems we really do need to own an `Options` instance at that point, in order to be able to pass it to `Configuration::from_options()`. To achieve that, I think we _have_ to clone the underlying `Options` instance at that point, unfortunately -- but I might be missing something

Reply:
You can maybe do a little better here with [`Arc::try_unwrap`](https://doc.rust-lang.org/std/sync/struct.Arc.html#method.try_unwrap):

```
diff --git a/crates/ruff/src/args.rs b/crates/ruff/src/args.rs
index 9d72c6168..4e74891a1 100644
--- a/crates/ruff/src/args.rs
+++ b/crates/ruff/src/args.rs
@@ -1,4 +1,4 @@
-use std::path::PathBuf;
+use std::{path::PathBuf, sync::Arc};
 
 use anyhow::anyhow;
 use clap::builder::{TypedValueParser, ValueParserFactory};
@@ -534,8 +534,10 @@ impl ConfigArgs {
         for option in config_options {
             match option {
                 ConfigOption::ConfigOverride(overridden_option) => {
+                    let overridden_option =
+                        Arc::try_unwrap(overridden_option).unwrap_or_else(|x| Options::clone(&x));
                     new.config_overrides = new.config_overrides.combine(
-                        Configuration::from_options(*overridden_option, &path_dedot::CWD)?,
+                        Configuration::from_options(overridden_option, &path_dedot::CWD)?,
                     );
                 }
                 ConfigOption::PathToConfigFile(path) => {
@@ -691,7 +693,7 @@ fn resolve_bool_arg(yes: bool, no: bool) -> Option<bool> {
 #[derive(Clone, Debug)]
 pub enum ConfigOption {
     PathToConfigFile(PathBuf),
-    ConfigOverride(Box<Options>),
+    ConfigOverride(Arc<Options>),
 }
 
 #[derive(Clone)]
@@ -722,7 +724,7 @@ impl TypedValueParser for ConfigOptionParser {
             return Ok(ConfigOption::PathToConfigFile(path_to_config_file));
         }
         if let Ok(option) = toml::from_str(value) {
-            Ok(ConfigOption::ConfigOverride(option))
+            Ok(ConfigOption::ConfigOverride(Arc::new(option)))
         } else {
             let mut new_error =
                 clap::Error::new(clap::error::ErrorKind::ValueValidation).with_cmd(cmd);
```

That way, there shouldn't be any unnecessary clones if nothing is actually cloning the `ConfigOption`. But it does still unfortunately require that `Options` implements `Clone`.

I don't see a simple way of avoiding the clone here without some other kind of bigger refactoring. I can think of some reasons why the clone is probably fine:

* This only applies when options are overridden on the CLI, which is probably not the common case.
* An `Options` in this case likely only has one non-`None` value, and so the actual heap allocation happening here is probably pretty small.

Reply:
With that said, it is unfortunate that this makes a bunch of non-`Clone` types `Clone`. That in turn will make it easier to clones to slip in in the future. But I don't see a simple way of avoiding that other than not using `Options` here. But that would require duplicating all of `Options` fields into another type (probably an enum of the options instead of a struct of the options). But then you'd have to impl `Clone` for that type, which would probably mean that you'd still need most of the added `Clone` impls in this PR.

The only way I see out of this is to abstain from using clap's `TypedValueParser`, which is the actual thing requiring a `Clone` implementation here. But I'm actually not familiar enough with Clap 4 to know how to do that off-hand. I took a quick look at Clap's API, and it doesn't look like there are any escape hatches. Which to me means the only way around this is to specifically opt out from Clap's "value parsing" and do it as a second step after getting the parsed args from Clap. Which will be supremely annoying I suspect.

The last idea is that `ConfigOption` be some kind of facsimile of the specific option being set from `Options`. Like for example, an `Option<(String, String)>`, with the first `String` being the key and the second `String` a value. Then only after Clap parsing would you parse that second `String` into an actual value. That would let you retain the use of Clap's value parsing machinery by side-stepping the need to involve `Options` into the trait impl. But... this is likely to prove annoying.

I think overall, on balance, doing what you did here is probably the right path even though it adds a bunch of `Clone` impls.

Reply:
Nice, thanks @BurntSushi! I switched to using an `Arc`, as you recommended.

I basically agree with everything you said above -- I tried to find a way to do it without cloning, but couldn't see a way without a large refactor of lots of other code. The fact that it's already done in various similar places also made me feel a little better about things:

https://github.com/astral-sh/ruff/blob/395bf3dc98a726eed64280211f06e92b3e309467/crates/ruff/src/args.rs#L666-L749

I'll leave this conversation "unresolved", though, in case anybody else wants to chime in on the approach here